dap-mode.el

```diff
@@ -683,14 +693,34 @@ thread exection but the server will log message."
   (run-hook-with-args 'dap-terminated-hook debug-session)
   (dap--refresh-breakpoints))
 
+(defun dap--output-buffer-format-with-category (category output)
+  "Formats a string suitable for printing to the output buffer using CATEGORY and OUTPUT."
+  (let ((message (format "%s: %s" category output)))
+    (if (string= (substring message -1) "\n")
+        message
+      (concat message "\n"))))
+
+(defun dap--output-buffer-format (output-body)
+  "Formats a string suitable for printing to the output buffer using an OUTPUT-BODY."
+  (if dap-label-output-buffer-category
+      (dap--output-buffer-format-with-category (gethash "category" output-body) (gethash "output" output-body))
+    (gethash "output" output-body)))
+
+(defun dap--print-to-output-buffer (debug-session str)
+  "Insert content from STR into the output buffer associated with DEBUG-SESSION."
+  (with-current-buffer (dap--debug-session-output-buffer debug-session)
+    (let ((win (get-buffer-window (current-buffer))))
```

Comment:
win is not used. Also, can you wrap it in save-excursion?

Reply:
wrapping this in `save-excursion` seems to break the scrolling functionality that was added from setting the `window-point-insertion-type`. I'm assuming this happens because when we set `save-excursion`, the point moves back to its original position after insertion, but I'm not 100% sure.

Reply:
The solution I've implemented is to check whether the output buffer is focused or not, and wrap `save-excursion` if it is focused. Let me know if this is suitable